docs/proposals/202005_query_logging.md

```diff
@@ -0,0 +1,257 @@
+
+**User specific configurations** : Currently we are going to provide the following user facing configurations for using the request logger - 
+* Use pre-defined adaptive logging policy.
+* The Adaptive Logging policy plans to include logging of only requests that return an error, or crosses a specific duration of time.
+* The level for high latency requests is `DEBUG`. Error, however would be logged at the `ERROR` level.
+* We would add in a flag to turn on/off request logging, and another flag to accept the log file path where we can log the details.
+* All the audit/adaptive logging can be logged into the respective files by passing the path of the files as a cli argument.
+* Request Logging would be turned off by default, and default logging would be in stdout. For logging all the requests, we would accept a flag for the enabling the same.
+* Currently the policy for request logging is hardcoded, may be extended as a user based policy later.
+
+**Types of Middlewares** : We are going to use the following middlewares for achieving the above objectives -
+
+* [`logging.WithDecider`](https://github.com/grpc-ecosystem/go-grpc-middleware/blob/v2/interceptors/logging/options.go#L51) - This `grpc.Option` would help us in deciding the logic whether a given request should be logged or not. This should help in logging specific / all queries
+* [`logging.WithLevels`](https://github.com/grpc-ecosystem/go-grpc-middleware/blob/v2/interceptors/logging/options.go#L58) - This `grpc.Option` would help us in fixating the level of query that we might want to log. So based on a query hitting a certain criteria, we can switch the levels of it
+* For using the request-id, we can store the request-id in the metadata of the `context`, and while logging, we can use it.
+* Same equivalent for the HTTP can be implemented for mirroring the logic of gRPC middlewares
+
+#### Request Logging
+
+Query Logging in Thanos would be implemented as a one logger, Request Logger. This is done in considering that Adaptive is a specific case of Audit Logging, and would make sense to keep the similar logical components together, rather than creating another distinction between them.
+We might however discuss the specific sub-cases of the request logging, namely adaptive and audit logging for enumerating the purpose of it.
+
+#### Use case 1: Audit Logging
+
+**Audit logging** is a kind of logging, where we log in every internal API requests. Typically, rather than logging in specific queries, we ought to log in every query that has been made until now.
+
+From a developer’s perspective, audit logs can keep you sane by giving you some insight into how a complex system arrived at its current state. Developers are typically fond of state diagrams and deterministic outcomes. An audit log that shows that a person or process applied a change can give developers valuable insight.
+
+#### Implementation:
+
+
+```go
```

Comment:
Those looks exactly like go kit logger, nothing special. In this form, we can skip it./

Reply:
Sure